File: src/main/java/org/funivan/lologa/iterator/Shuffled.java
package org.funivan.lologa.iterator;

import org.cactoos.scalar.StickyScalar;
import org.cactoos.scalar.SyncScalar;
import org.cactoos.scalar.UncheckedScalar;

import java.util.Collections;
import java.util.Iterator;
import java.util.LinkedList;

public final class Shuffled<T> implements Iterator<T> {
    private final UncheckedScalar<Iterator<T>> scalar;

    public Shuffled(Iterator<T> iterator) {
        this.scalar = new UncheckedScalar(
            new SyncScalar(
                new StickyScalar(() -> {
                    LinkedList items = new LinkedList();
                    while (iterator.hasNext()) {
                        items.add(iterator.next());
                    }
                    Collections.shuffle(items);
                    return items.iterator();
                })
            )
        );
    }

    public boolean hasNext() {
        return this.scalar.value().hasNext();
    }

    public T next() {
        return this.scalar.value().next();
    }
}

File: src/main/java/org/funivan/lologa/iterator/Shuffled.kt
package org.funivan.lologa.iterator

import org.cactoos.list.ListOf
import java.util.*

class Shuffled<T>(val origin: Iterator<T>) : Iterator<T> {
    private var iter: Iterator<T> = ListOf<T>().iterator()
    private var shuffled = false;

    private fun shuffledIterator(): Iterator<T> {
        if (!this.shuffled) {
            var a = mutableListOf<T>()
            while (this.origin.hasNext()) {
                a.add(this.origin.next())
            }
            Collections.shuffle(a);
            this.iter = a.iterator();
            this.shuffled = true;
        }
        return this.iter
    }

    override fun hasNext(): Boolean {
        return shuffledIterator().hasNext()
    }

    override fun next(): T {
        return this.shuffledIterator().next()
    }
}

